controllers/api/webhook.go

```diff
@@ -0,0 +1,96 @@
+package api
+
+import (
+	"encoding/json"
+	"net/http"
+	"strconv"
+
+	log "github.com/gophish/gophish/logger"
+	"github.com/gophish/gophish/models"
+	"github.com/gophish/gophish/webhook"
+	"github.com/gorilla/mux"
+)
+
+// Webhooks returns a list of webhooks, both active and disabled
+func (as *Server) Webhooks(w http.ResponseWriter, r *http.Request) {
+	switch {
+	case r.Method == "GET":
+		whs, err := models.GetWebhooks()
+		if err != nil {
+			log.Error(err)
+			JSONResponse(w, models.Response{Success: false, Message: err.Error()}, http.StatusInternalServerError)
+			return
+		}
+		JSONResponse(w, whs, http.StatusOK)
+
+	case r.Method == "POST":
+		wh := models.Webhook{}
+		err := json.NewDecoder(r.Body).Decode(&wh)
+		if err != nil {
+			JSONResponse(w, models.Response{Success: false, Message: "Invalid JSON structure"}, http.StatusBadRequest)
+			return
+		}
+		err = models.PostWebhook(&wh)
+		if err != nil {
+			JSONResponse(w, models.Response{Success: false, Message: err.Error()}, http.StatusBadRequest)
+			return
+		}
+		JSONResponse(w, wh, http.StatusCreated)
+	}
+}
+
+// Webhook returns details of a single webhook specified by "id" parameter
+func (as *Server) Webhook(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id, _ := strconv.ParseInt(vars["id"], 0, 64)
+	wh, err := models.GetWebhook(id)
+	if err != nil {
+		JSONResponse(w, models.Response{Success: false, Message: "Webhook not found"}, http.StatusNotFound)
+		return
+	}
+	switch {
+	case r.Method == "GET":
+		JSONResponse(w, wh, http.StatusOK)
+
+	case r.Method == "DELETE":
+		err = models.DeleteWebhook(id)
+		if err != nil {
+			JSONResponse(w, models.Response{Success: false, Message: err.Error()}, http.StatusInternalServerError)
+			return
+		}
+		log.Infof("Deleted webhook with id: %d", id)
+		JSONResponse(w, models.Response{Success: true, Message: "Webhook deleted Successfully!"}, http.StatusOK)
+
+	case r.Method == "PUT":
+		wh2 := models.Webhook{}
+		err = json.NewDecoder(r.Body).Decode(&wh2)
+		wh2.Id = id
+		err = models.PutWebhook(&wh2)
+		if err != nil {
+			JSONResponse(w, models.Response{Success: false, Message: err.Error()}, http.StatusBadRequest)
+			return
+		}
+		JSONResponse(w, wh2, http.StatusOK)
+	}
+}
+
+// ValidateWebhook makes an HTTP request to a specified remote url to ensure that it's valid.
+func (as *Server) ValidateWebhook(w http.ResponseWriter, r *http.Request) {
+	switch {
+	case r.Method == "POST":
+		vars := mux.Vars(r)
+		id, _ := strconv.ParseInt(vars["id"], 0, 64)
+		wh, err := models.GetWebhook(id)
+		if err != nil {
+			log.Error(err)
+			JSONResponse(w, models.Response{Success: false, Message: err.Error()}, http.StatusInternalServerError)
+			return
+		}
+		err = webhook.Send(webhook.EndPoint{URL: wh.URL, Secret: wh.Secret}, "")
+		if err == nil {
```

Comment:
How will flipping the conditionals clean up the code? We're expect the condition without an error first. And that's what I'm calling here first. And there have to exist 2 conditions here. 

```
		err = webhook.Send(webhook.EndPoint{URL: wh.URL, Secret: wh.Secret}, "")
		if err == nil {
			JSONResponse(w, wh, http.StatusOK)
		} else {
			JSONResponse(w, models.Response{Success: false, Message: err.Error()}, http.StatusBadRequest)
		}
```


How do you say we could do this?

```
		err = webhook.Send(webhook.EndPoint{URL: wh.URL, Secret: wh.Secret}, "")
		if err != nil {
			JSONResponse(w, models.Response{Success: false, Message: err.Error()}, http.StatusBadRequest)
		} 
```

Return nothing in case of "no error"?

Reply:
My recommendation is really just encouraging the pattern we use everywhere else in the code, which is also what Go itself recommends. That is, a pattern like:

```
if err != nil {
    handle the error
    return
}
continue execution
```

The reason this is encouraged is that it keeps the nesting to a minimum so we don't wind up with things like:

```
if err == nil {
    big block of code that happens if things go as planned
    if err == nil {
        another big block of code
        etc, etc
    } else {
        handle the error
    }
} else {
    handle the error
}
```

So in my proposal we're handling the error first, then if the error was nil, you'll find the `JSONResponse(w, wh, http.StatusOK)` at the end of the suggested code- this would be the default thing to run if no errors occurred. That's the way all of the other HTTP handlers work- this is the only one that has a different pattern when it doesn't need to.

Reply:
Agree. Fixed